pulp_ansible/app/custom_fields.py

```diff
@@ -0,0 +1,209 @@
+
+from django.utils.translation import gettext_lazy as _
+from django.core.exceptions import BadRequest
+
+from drf_spectacular.utils import extend_schema_field
+
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError
+
+from pulpcore.plugin.models.role import Role
+
+from pulpcore.plugin.util import (
+    assign_role,
+    remove_role,
+    get_groups_with_perms_attached_roles,
+    get_perms_for_model,
+)
+from pulpcore.plugin.models import Group
+
+
+class RelatedFieldsBaseSerializer(serializers.Serializer):
+    """
+    Serializer only returns fields specified in 'include_related' query param.
+
+    This allows for fields that require more database queries to be optionally
+    included in API responses, which lowers the load on the backend. This is
+    intended as a way to include extra data in list views.
+
+    Usage:
+
+    This functions the same as DRF's base `serializers.Serializer` class with the
+    exception that it will only return fields specified in the `?include_related=`
+    query parameter.
+
+    Example:
+
+    MySerializer(RelatedFieldsBaseSerializer):
+        foo = CharField()
+        bar = CharField()
+
+    MySerializer will return:
+
+    {"foo": None} when called with `?include_related=foo` and {"foo": None, "bar" None}
+    when called with `?include_related=foo&include_related=bar`.
+    """
+
+    def __init__(self, *args, **kwargs):
+        # This should only be included as a sub serializer and shouldn't be used for
+        # updating objects, so set read_only to true
+        kwargs["read_only"] = True
+        return super().__init__(*args, **kwargs)
+
+    def to_representation(self, instance):
+        result = OrderedDict()
+        fields = self._readable_fields
+        request = self.context.get("request", None)
+        if request:
+            # TODO: Figure out how to make `include_related` show up in the open API spec
+            include_fields = request.GET.getlist("include_related")
+
+            if len(include_fields) > 0:
+                for field in fields:
+                    if field.field_name in include_fields:
+                        result[field.field_name] = field.to_representation(instance)
+        else:
+            # When there is no request present, it usually means that the serializer is
+            # being inspected by DRF spectacular to generate the open API spec. In that
+            # case, this should act like a normal serializer.
+            return super().to_representation(instance)
+
+        return result
+
+
+class MyPermissionsField(serializers.Serializer):
+    def to_representation(self, original_obj):
+        request = self.context.get("request", None)
+        if request is None:
+            return []
+        user = request.user
+
+        if original_obj._meta.proxy:
+            obj = original_obj._meta.concrete_model.objects.get(pk=original_obj.pk)
+        else:
+            obj = original_obj
```

Comment:
I'm pretty sure, in Pulp we use permissions on the proxy model is that different in Galaxy?

(I think the Group is the only proxy model we use, and we needed that for additional permissions).